src/xml/builder.cr

```diff
@@ -44,11 +44,14 @@ struct XML::Builder
 
   # Emits the start of an element.
   def start_element(name : String) : Nil
-    call StartElement, string_to_unsafe(name)
+    element_name = string_to_unsafe name
+    raise XML::Error.new("Invalid element name: '#{name}'", __LINE__) if LibXML.xmlValidateNameValue(element_name).zero?
+    call StartElement, element_name
   end
 
   # Emits the start of an element with namespace info.
   def start_element(prefix : String?, name : String, namespace_uri : String?) : Nil
+    raise XML::Error.new("Invalid element name: '#{prefix}#{name}'", __LINE__) if LibXML.xmlValidateNameValue(string_to_unsafe("#{prefix}#{name}")).zero?
```

Comment:
Shouldn't `prefix` and `name` be validated separately? Or at least, would it be the same to do that? Then we don't need a string interpolation, plus we could say "Invalid prefix" or "Invalid element name".

Reply:
Ah, good call.  I'll get that fixed.

Reply:
Thank you! But now `string_to_unsafe` is called twice for each `name` and `prefix`, and that does a null byte checking so now that is done twice. So an optional thing to do is to call that first and then pass those things to `check_valid...`, though I can see it starts to become a bit annoying. Sorry! So feel free to ignore this comment, eventually we can optimize that.